pkg/backup/backup_compaction_test.go

```diff
@@ -0,0 +1,291 @@
+		)
+	})
+
+	t.Run("create indexes", func(t *testing.T) {
+		db.Exec(t, "CREATE TABLE foo (a INT, b INT)")
+		defer func() {
+			db.Exec(t, "DROP TABLE foo")
+		}()
+		for _, stmt := range insertRows(10, "foo") {
+			db.Exec(t, stmt)
+		}
+		db.Exec(t, fullBackupStmt)
+		runQueriesAndBackup(
+			t, db, "'nodelocal://1/backup'",
+			insertRows(20, "foo"), []string{}, false,
+		)
+		runQueriesAndBackup(
+			t, db, "'nodelocal://1/backup'",
+			insertRows(10, "foo"), []string{}, false,
+		)
+		runQueriesAndBackup(
+			t, db, "'nodelocal://1/backup'",
+			[]string{"CREATE INDEX ON foo (a)"}, []string{}, true,
+		)
+		var numIndexes, restoredNumIndexes int
+		db.QueryRow(t, "SELECT count(*) FROM [SHOW INDEXES FROM foo]").Scan(&numIndexes)
+		db.Exec(t, "DROP TABLE foo")
+		db.Exec(t, "RESTORE TABLE foo FROM LATEST IN 'nodelocal://1/backup'")
+		db.QueryRow(t, "SELECT count(*) FROM [SHOW INDEXES FROM foo]").Scan(&restoredNumIndexes)
+		require.Equal(t, numIndexes, restoredNumIndexes)
+	})
+
+	// TODO (kev-cao): Once range keys are supported by the compaction
+	// iterator, add tests for dropped tables/indexes.
+}
+
+func TestBackupCompactionMultiRegion(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	tempDir, tempDirCleanup := testutils.TempDir(t)
+	defer tempDirCleanup()
+	_, db, cleanupDB := backupRestoreTestSetupEmpty(
+		t, multiNode, tempDir, InitManualReplication,
+		base.TestClusterArgs{
+			ServerArgsPerNode: map[int]base.TestServerArgs{
+				0: {
+					Locality: roachpb.Locality{Tiers: []roachpb.Tier{
+						{Key: "region", Value: "west"},
+						{Key: "az", Value: "az1"},
+						{Key: "dc", Value: "dc1"},
+					}},
+				},
+				1: {
+					Locality: roachpb.Locality{Tiers: []roachpb.Tier{
+						{Key: "region", Value: "east"},
+						{Key: "az", Value: "az1"},
+						{Key: "dc", Value: "dc2"},
+					}},
+				},
+				2: {
+					Locality: roachpb.Locality{Tiers: []roachpb.Tier{
+						{Key: "region", Value: "east"},
+						{Key: "az", Value: "az2"},
+						{Key: "dc", Value: "dc3"},
+					}},
+				},
+			},
+		},
+	)
+	defer cleanupDB()
+	collectionURIs := strings.Join([]string{
+		fmt.Sprintf(
+			"'nodelocal://1/backup?COCKROACH_LOCALITY=%s'",
+			url.QueryEscape("default"),
+		),
+		fmt.Sprintf(
+			"'nodelocal://2/backup?COCKROACH_LOCALITY=%s'",
+			url.QueryEscape("dc=dc2"),
+		),
+		fmt.Sprintf(
+			"'nodelocal://3/backup?COCKROACH_LOCALITY=%s'",
+			url.QueryEscape("region=west"),
+		),
+	}, ", ")
+	db.Exec(t, "CREATE TABLE foo (a INT, b INT)")
+	db.Exec(t, fmt.Sprintf("BACKUP INTO (%s)", collectionURIs))
+	createInsertQueries := func(count int) []string {
+		insertQueries := make([]string, 0, count)
+		for i := 0; i < count; i++ {
+			a, b := rand.IntN(1000), rand.IntN(100)
+			insertQueries = append(insertQueries, fmt.Sprintf("INSERT INTO %s VALUES (%d, %d)", "foo", a, b))
+		}
+		return insertQueries
+	}
+	runQueriesAndBackup(t, db, collectionURIs, createInsertQueries(10), []string{}, false)
```

Comment:
why is it necessary to run all these insert  and update queries to test that we handle locality aware backups properly?

Reply:
Updated the code so that it's one insert statement instead of several, but to answer your question, I wanted to have some deltas in each backup to ensure that compaction was doing something meaningful so that it could be validated.